src/coreComponents/mesh/MeshLevel.cpp

```diff
@@ -121,35 +119,43 @@ MeshLevel::MeshLevel( string const & name,
   m_faceManager->faceNormal() = source.m_faceManager->faceNormal();
 
   // Faces
-//  ArrayOfArraysView< localIndex const > const facesToNodesMapSource = m_faceManager->nodeList().toViewConst();
   ArrayOfArrays< localIndex > & faceToNodeMapNew = m_faceManager->nodeList();
   ArrayOfArraysView< localIndex const > const & facesToEdges = m_faceManager->edgeList().toViewConst();
-  localIndex const estimatedNumNodesPerFace = pow( order+1, 2 );
-  faceToNodeMapNew.resize( faceToNodeMapNew.size(), estimatedNumNodesPerFace );
+  localIndex const numNodesPerFace = pow( order+1, 2 );
 
 
   // add the number of non-edge face nodes
   localIndex numInternalFaceNodes = 0;
+  localIndex const numNonEdgeNodesPerFace = pow( order-1, 2 );
   for( localIndex kf=0; kf<m_faceManager->size(); ++kf )
   {
     localIndex const numEdgesPerFace = facesToEdges.sizeOfArray( kf );
-//    localIndex const numVertexNodesPerFace = facesToNodesMapSource.sizeOfArray( kf );
-//    localIndex const numEdgeNodesPerFace = numVertexNodesPerFace + numEdgesPerFace * numNonVertexNodesPerEdge;
-
     if( numEdgesPerFace==4 )
-    {
-      localIndex const numNonEdgeNodesPerFace = pow( order-1, 2 );
       numInternalFaceNodes += numNonEdgeNodesPerFace;
-
-      localIndex const numNodesPerFace = pow( order+1, 2 );
-      faceToNodeMapNew.resizeArray( kf, numNodesPerFace );
-    }
     else
     {
       GEOSX_ERROR( "need more support for face geometry" );
     }
   }
 
+  // number of elements in each row of the map as capacity
+  array1d< localIndex > counts( faceToNodeMapNew.size());
+  counts.setValues< parallelHostPolicy >( numNodesPerFace );
+
+  //  reconstructs the faceToNodeMap with the provided capacity in counts
+  faceToNodeMapNew.resizeFromCapacities< parallelHostPolicy >( faceToNodeMapNew.size(), counts.data() );
+
+  // setup initial values of the faceToNodeMap using emplaceBackAtomic
+  forAll< parallelHostPolicy >( numNodesPerFace, [faceToNodeMapNew = faceToNodeMapNew.toView()] ( localIndex faceIndex )
+  {
+    for( faceIndex =0; faceIndex < faceToNodeMapNew.size(); faceIndex++ )
+    {
+      faceToNodeMapNew.emplaceBackAtomic< AtomicPolicy< parallelHostPolicy > >( faceIndex, 0 );
+    }
+
+  } );
```

Comment:
Thanks for the suggestion! I choose "emplaceBackAtomic" here to allow multiple threads to append to a single inner array in a thread safe manner. To be more specific, since parallelHostPolicy is utilized in the inner array for appending in parallel, I think atomic functions would be needed to provide low-level locking mechanisms for synchronizing accesses. I would be glad to discuss on this in more details to make sure this is the best way to do it :).

And yes you are right, for filling the dummy values in to the arrays we should be filling with -1 rather than 0. Good point!

Reply:
@klevzoff Are you OK merging this or do you think it's worth modifying?

Reply:
Sorry for disengaging from the conversation earlier. I would prefer to make my proposed change, if we agree that it achieves an equivalent result (the proposed code was not tested, so it would take a build/run to verify). Reasons being:
  * more parallelism (in particular, using more than 4 threads)
  * no use of atomics (which are relatively slow)
  * better memory/cache locality (each thread writes to a contiguous chunk of the map)

I don't think the difference here would be particularly noticeable (this is a trivial memory-bound operation anyway), but as a principle we should try to avoid atomics and coarse-grained parallelism where we can. As always, one way to change my mind is with benchmarks :) but this one is not worth the time...

Reply:
IIUC current impl is a parallel loop which unique action is an atomic?
Removing this atomic sounds like a good move indeed.

Reply:
Sure, good idea. I will benchmark it tomorrow and make the proposed change if performance doesn’t change. :)

Reply:
This `emplace_back()` strategy  only works if enough space was allocated for each face...right? For instance, if there were not enough `estimatedNumNodesPerFace` (now deleted and hard coded) for all faces, then a reallocation might be triggered, which would cause a race?

Reply:
That's right, both versions of this loop rely on previous allocation being exact. Neither will actually reallocate (since we're calling `emplaceBack` on a view), but trigger an assertion.

Reply:
Yes I agree. In this case the race condition wouldn't be a concern. @klevzoff thanks for the suggestion and clarification, I just pushed the proposed change. 👍